pre_commit/languages/lua.py

```diff
@@ -0,0 +1,137 @@
+import contextlib
+import os
+import re
+from typing import Generator
+from typing import Sequence
+from typing import Tuple
+
+from pre_commit.envcontext import envcontext
+from pre_commit.envcontext import PatchesT
+from pre_commit.envcontext import Var
+from pre_commit.hook import Hook
+from pre_commit.languages import helpers
+from pre_commit.prefix import Prefix
+from pre_commit.util import clean_path_on_failure
+from pre_commit.util import cmd_output
+
+ENVIRONMENT_DIR = 'lua_env'
+get_default_version = helpers.basic_get_default_version
+healthy = helpers.basic_healthy
+
+
+def _find_lua() -> str:  # pragma: win32 no cover
+    """Find a lua executable.
+
+    Lua doesn't always have a plain `lua` executable.
+    Some OS vendors will ship the binary as `lua#.#` (e.g., lua5.3)
+    so discovery is needed to find a valid executable.
+    """
+    paths = os.environ.get('PATH', '').split(os.pathsep)
+    possible_luas = []
+    for path in paths:
+        try:
+            candidate_executables = os.listdir(path)
+        except FileNotFoundError:
+            # Invalid path on PATH.
+            continue
+
+        for candidate in candidate_executables:
+            if candidate == 'lua':
+                # Found one. Bail early.
+                return os.path.join(path, candidate)
+            elif candidate.startswith('lua'):
+                possible_luas.append((path, candidate))
```

Comment:
this needs to check that they are executable and a a file -- otherwise this will discover non-executables

one way you could do this ~slightly easier would be to collect all the possible basenames and then use `parse_shebang.find_executable` to return the first non-None one -- that'd reuse most of the `PATH` parsing for you

Reply:
I'm not sure how to make it work with `find_executable`. I like the idea of reusing the `PATH` parsing, but I don't think I'd be able to construct the possible basenames to check against. The problem that I see currently is that the information about the version isn't available.

For instance, the Ubuntu packaging scenario makes building basenames difficult. The user may not provide a version so pre-commit would be set to use the default. Since the Ubuntu package doesn't provide a `lua` executable and only provides something like `lua-5.3`, there's no way to guess the 5.3 value. The basenames could be `['lua']` in this case and, on Ubuntu, pre-commit would fail to find the executable.

As far as I can tell, sniffing through all the executables with the regex is the only way I can see to get the lua executable in that scenario.

Any ideas on how to work around that scenario?

Reply:
basically something like this:

```python
choices = {'lua'}
for path in os.environ.get('PATH', '').split(':'):
    try:
        candidates = os.listdir(path)
    except OSError:
        continue
    else:
        for candidate in candidates:
            if re.match('^lua-\d+\.\d+(\.exe)?$', candidate):
                choices.add(candidate)

found_exes = [
    exe for exe in choices
    if find_executable(exe)
]
# ... pick the one that's newest or is `lua`
```

Reply:
Ah, I see what you mean. I got the impression that you thought their was a way to do this by iterating over the PATH paths one time.

I'll take your approach. I think I can simplify the regex too because the `.exe` extension handling is factored in to `find_executable`. Thanks for the code sketch.